lib/backend/dir/impl.go

```diff
@@ -82,318 +82,306 @@ func New(params backend.Params) (backend.Backend, error) {
 		return nil, trace.BadParameter("filesystem backend: 'path' is not set")
 	}
 
+	// Ensure that the path to the root directory exists.
+	err := os.MkdirAll(rootDir, defaultDirMode)
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
 	bk := &Backend{
-		RootDir:       rootDir,
 		InternalClock: clockwork.NewRealClock(),
-		Entry: log.WithFields(log.Fields{
+		rootDir:       rootDir,
+		log: logrus.WithFields(logrus.Fields{
 			trace.Component: "backend:dir",
-			trace.ComponentFields: log.Fields{
+			trace.ComponentFields: logrus.Fields{
 				"dir": rootDir,
 			},
 		}),
 	}
 
-	locksDir := path.Join(bk.RootDir, locksBucket)
-	if err := os.MkdirAll(locksDir, defaultDirMode); err != nil {
-		return nil, trace.ConvertSystemError(err)
+	// DELETE IN: 2.8.0
+	// Migrate data to new flat keyspace backend.
+	err = migrate(rootDir, bk)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
 
 	// Wrap the backend in a input sanitizer and return it.
 	return backend.NewSanitizer(bk), nil
 }
 
-// GetItems is a function that returns keys in batch
-func (bk *Backend) GetItems(bucket []string) ([]backend.Item, error) {
-	keys, err := bk.GetKeys(bucket)
+// Close releases the resources taken up the backend.
+func (bk *Backend) Close() error {
+	return nil
+}
+
+// GetKeys returns a list of keys for a given bucket.
+func (bk *Backend) GetKeys(bucket []string) ([]string, error) {
+	// Get all the key/value pairs for this bucket.
+	items, err := bk.GetItems(bucket)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	var items []backend.Item
-	for _, key := range keys {
-		v, err := bk.GetVal(bucket, key)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		items = append(items, backend.Item{Key: key, Value: v})
+
+	// Return only the keys, the keys are already sorted by GetItems.
+	keys := make([]string, len(items))
+	for i, e := range items {
+		keys[i] = e.Key
 	}
-	return items, nil
+
+	return keys, nil
 }
 
-// GetKeys returns a list of keys for a given path
-func (bk *Backend) GetKeys(bucket []string) ([]string, error) {
-	files, err := ioutil.ReadDir(path.Join(bk.RootDir, path.Join(bucket...)))
+// GetItems returns all items (key/value pairs) in a given bucket.
+func (bk *Backend) GetItems(bucket []string) ([]backend.Item, error) {
+	var out []backend.Item
+
+	// Get a list of all buckets in the backend.
+	files, err := ioutil.ReadDir(path.Join(bk.rootDir))
```

Comment:
you know here in advance the list of buckets files to open, why do a full scan?

Reply:
At this point, the actual bucket to open is not known. Take for example a backend with the following items in it:

```
/roles/admin/params
/roles/dev/params
/namespaces/default/nodes/123
```

If bucket is `/namespaces/default/nodes/123`, then GetItems don't not need to do a full directory scan and can return the exact bucket (or return `trace.NotFound` if it does not exist).

However if the caller asks for `/roles` (which is what `GetRoles` does), that's when the directory scan needs to occur perform a prefix match then extract the first matching bucket and return it as a key.